src/devinterp/slt/norms.py

```diff
@@ -51,4 +52,28 @@ def _flatten_grads(self, model: nn.Module):
         grads = [param.grad.detach().clone() for param in model.parameters()]
         grads = [torch.flatten(grad) for grad in grads]
         grads = torch.cat(grads)
-        return grads
\ No newline at end of file
+        return grads
+    
+
+class NoiseNorm(ChainCallback):
+    def __init__(self, num_chains, num_draws, device='cpu'):
+        self.noise_norms = torch.tensor((num_chains, num_draws), dtype=torch.float32).to(device)
+        self.device = device
+    
+    def __call__(self, chain: int, draw: int, optimizer: SGLD):
+        self.update(chain, draw, optimizer)
+    
+    def update(self, chain: int, draw: int, optimizer: SGLD):
+        noise = self._flatten_noise(optimizer)
+        self.noise_norms[chain, draw] = torch.linalg.vector_norm(noise, ord=2)
+
+    def sample(self):
+        return {
+            "noise_norm/trace": self.noise_norms.cpu().numpy()
+        }
+    
+    def _flatten_noise(self, optimizer: SGLD):
```

Comment:
I would avoid the flatten noise approach because it has more memory overhead. You can compute the norms of each noise vector one at a time and then just add them up instead of having to create a new tensor

Reply:
Maybe I'm not following the norm we want to use (or what we're trying to norm), if we're using L2 norm then these two are not equivalent implementations. See here: https://gist.github.com/georgeyw/d5fa3b94ed0d93d1c1148075d80b3776

Reply:
They also don't always agree on ordinality: suppose you have a simple model with three sets of parameters, each being one tensor (so shape (3, 1)), where noise1 is 1, 0, 0 and noise2 is 0.5, 0.5, 0.5 for each layer respectively.

(1, 0, 0) has a noise norm of 1 in either implementation

(0.5, 0.5, 0.5) has a noise norm of ~0.87 in one implementation and 1.5 in the other